### PR TITLE
Staging/next major

### PR DESCRIPTION
**Pull request by netanelc305**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,7 +66,21 @@ Use the attach sub-command in order to start an LLDB shell attached to given pro
 hilda attach [-p pid] [-n process-name]
 ```
 
-After attaching, simply execute `hilda` command to enter the hilda shell.
+### Launch mode
```
Comment (doronz88) [inline README.md:69]: update the TOC with these headings

File: hilda/cli.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+@click.group()
+def cli():
+    pass
+
+
+def common_options(func: Callable) -> Callable:
+    """ Decorator to add common options to commands """
+
+    @click.option('-f', '--startup_files', multiple=True, default=None, help='Files to run on start')
```
Comment (doronz88) [inline hilda/cli.py:17]: lets refactor this into a single line decorator

File: hilda/cli.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    @click.pass_context
+    def new_func(ctx, *args, **kwargs):
+        return ctx.invoke(func, *args, **kwargs)
+
+    return new_func
+
+
+def parse_envp(ctx, param, value):
```
Comment (doronz88) [inline hilda/cli.py:25]: add typings

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: hilda/hilda_client.py [hunk 4]
```diff
@@ -144,7 +164,7 @@ def lsof(self) -> dict:
         # convert FDs into int
         return {int(k): v for k, v in result.items()}
 
-    def bt(self, should_print=True, depth: Optional[int] = None) -> List:
+    def bt(self, should_print=False, depth: Optional[int] = None) -> List:
```
Comment (doronz88) [inline hilda/hilda_client.py:167]:
```suggestion
    def bt(self, should_print: bool = False, depth: Optional[int] = None) -> List[str, SBFrame]:
```

File: hilda/hilda_client.py [hunk 5]
```diff
@@ -302,7 +326,7 @@ def peek_str(self, address: Symbol) -> str:
         """
         return address.po('char *')[1:-1]  # strip the ""
 
-    def stop(self):
+    def stop(self, *args):
```
Comment (doronz88) [inline hilda/hilda_client.py:329]:
```suggestion
    def stop(self, *args) -> None:
```

File: hilda/hilda_client.py [hunk 6]
```diff
@@ -313,8 +337,10 @@ def stop(self):
 
         if not self.process.Stop().Success():
             self.log_critical('failed to stop process')
+        else:
+            self.log_info('Process Stopped')
 
-    def cont(self):
+    def cont(self, *args):
```
Comment (doronz88) [inline hilda/hilda_client.py:343]:
```suggestion
    def cont(self, *args) -> None:
```

File: hilda/hilda_client.py [hunk 7]
```diff
@@ -338,8 +366,11 @@ def detach(self):
         """
         if not self.process.Detach().Success():
             self.log_critical('failed to detach')
+        else:
+            self.log_info('Process Detached')
 
-    def disass(self, address, buf, flavor='intel', should_print=True) -> lldb.SBInstructionList:
+    @stop_is_needed
+    def disass(self, address, buf, flavor='intel', should_print=False) -> lldb.SBInstructionList:
```
Comment (doronz88) [inline hilda/hilda_client.py:373]:
```suggestion
    def disass(self, address: int, buf: bytes, flavor: str = 'intel', should_print: bool = False) -> lldb.SBInstructionList:
```

File: hilda/hilda_client.py [hunk 8]
```diff
@@ -1016,28 +1020,31 @@ def add_lldb_symbol(self, symbol: lldb.SBSymbol) -> Symbol:
 
         return value
 
-    def interact(self, additional_namespace: typing.Mapping = None) -> None:
+    def interact(self, additional_namespace: typing.Mapping = None, startup_files: Optional[List[str]] = None) -> None:
```
Comment (doronz88) [inline hilda/hilda_client.py:1023]:
```suggestion
    def interact(self, additional_namespace: Optional[typing.Mapping] = None, startup_files: Optional[List[str]] = None) -> None:
```

File: hilda/ipython_extensions/events.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import builtins
+
+import lldb
+
+from hilda.exceptions import SymbolAbsentError
+from hilda.symbols_jar import SymbolsJar
+
+
+class HIEvents:
+    def __init__(self, ip):
```
Comment (doronz88) [inline hilda/ipython_extensions/events.py:11]: add typing

File: hilda/ipython_extensions/events.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        self.hilda_client = self.shell.user_ns['p']
+
+    def pre_run_cell(self, info):
+        """
+        Enable lazy loading for symbols
+        :param info: IPython's CellInfo object
+        """
+        if info.raw_cell[0] in ['!', '%'] or info.raw_cell.endswith('?'):
+            return
+        #
```
Comment (doronz88) [inline hilda/ipython_extensions/events.py:22]: remove

File: hilda/ipython_extensions/keybindings.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+def load_ipython_extension(ipython):
+    def register_keybindings():
+        hilda = ipython.user_ns.get('p')
+        if not hilda:
+            print("Hilda instance not found in IPython namespace.")
+            return
+
+        insert_mode = ViInsertMode() | EmacsInsertMode()
+
+        if hasattr(ipython, 'pt_app'):
```
Comment (doronz88) [inline hilda/ipython_extensions/keybindings.py:15]: on which case doesn't it have this attribute?

File: hilda/ipython_extensions/keybindings.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+from prompt_toolkit.enums import DEFAULT_BUFFER
+from prompt_toolkit.filters import EmacsInsertMode, HasFocus, HasSelection, ViInsertMode
+from prompt_toolkit.keys import Keys
+
+
+def load_ipython_extension(ipython):
+    def register_keybindings():
+        hilda = ipython.user_ns.get('p')
+        if not hilda:
+            print("Hilda instance not found in IPython namespace.")
```
Comment (doronz88) [inline hilda/ipython_extensions/keybindings.py:10]: raise instead

File: hilda/ipython_extensions/keybindings.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+            return
+
+        insert_mode = ViInsertMode() | EmacsInsertMode()
+
+        if hasattr(ipython, 'pt_app'):
+            registry = ipython.pt_app.key_bindings
+
+            registry.add_binding(Keys.F7, filter=(HasFocus(DEFAULT_BUFFER) & ~HasSelection() & insert_mode))(
```
Comment (doronz88) [inline hilda/ipython_extensions/keybindings.py:18]: put this filter in its own variable

File: hilda/launch_lldb.py [hunk 14]
```diff
@@ -1,85 +1,195 @@
...
+        process = self.target.AttachToProcessWithName(self.listener, self.proc_name, self.wait_for, self.error)
+        if not self.error.Success():
+            raise LLDBException(self.error.description)
+        return process
+
+
+class LLDBLaunch(LLDBListenerThread):
+
+    def __init__(self, exec_path: str, argv: Optional[List[str]] = None, envp: Optional[List[str]] = None,
+                 stdin: Optional[str] = None,
```
Comment (doronz88) [inline hilda/launch_lldb.py:126]: remove line break

File: hilda/launch_lldb.py [hunk 15]
```diff
@@ -1,85 +1,195 @@
...
-    attach(name=name, pid=pid)
+def execute(cmd: str) -> int:
+    logging.debug(f'executing: {cmd}')
+    return os.system(cmd)
 
 
-if __name__ == '__main__':
-    disable_logs()
-    cli()
+class LLDBListenerThread(Thread):
```
Comment (doronz88) [inline hilda/launch_lldb.py:32]: make it an abstract class and make the two `_create_*` methods into `@abstractmethod`

File: hilda/launch_lldb.py [hunk 16]
```diff
@@ -1,85 +1,195 @@
...
+        raise NotImplementedError()
+
+    def _create_process(self) -> lldb.SBProcess:
+        raise NotImplementedError()
+
+    def run(self):
+        event = lldb.SBEvent()
+        last_state = lldb.eStateStopped
+        while not self.should_quit:
+            if self.listener.WaitForEvent(1, event):
```
Comment (doronz88) [inline hilda/launch_lldb.py:55]: mvoe the `1` into its own const `TIMEOUT`

File: hilda/launch_lldb.py [hunk 17]
```diff
@@ -1,85 +1,195 @@
...
+        self.url_connect = f'connect://{address}:{port}'
+        super().__init__()
+
+    def _create_target(self) -> lldb.SBTarget:
+        return self.debugger.CreateTarget('')
+
+    def _create_process(self) -> lldb.SBProcess:
+        logger.debug(f'Connecting to "{self.url_connect}"')
+        process = self.target.ConnectRemote(self.listener, self.url_connect, None, self.error)
+        if not self.error.Success():
```
Comment (doronz88) [inline hilda/launch_lldb.py:83]: move into its own method

File: hilda/launch_lldb.py [hunk 18]
```diff
@@ -1,85 +1,195 @@
...
+    """ Connect to remote process """
+    try:
+        lldb_t = LLDBRemote(hostname, port)
+        lldb_t.start()
+        hilda(lldb_t.debugger, startup_files)
+    except LLDBException as e:
+        logger.warning(e.message)
+
+
+def attach(name: Optional[str] = None, pid: Optional[int] = None, wait_for=False,
```
Comment (doronz88) [inline hilda/launch_lldb.py:167]:
```suggestion
def attach(name: Optional[str] = None, pid: Optional[int] = None, wait_for: bool = False,
```

File: hilda/launch_lldb.py [hunk 19]
```diff
@@ -1,85 +1,195 @@
...
-        commands.append(f'process attach -p {pid}')
-    else:
-        print('missing either process name or pid for attaching')
-        return
-
-    commands.append(f'command script import {os.path.join(Path(__file__).resolve().parent, "lldb_entrypoint.py")}')
-    if rc_script is not None:
-        commands.append(f'command script import {rc_script}')
-    commands = '\n'.join(commands)
+coloredlogs.install(level=logging.DEBUG)
```
Comment (doronz88) [inline hilda/launch_lldb.py:12]: should only be set on `cli.py` so only CLI will affect the application logging

File: hilda/cli.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+           flags: Optional[int] = 0, stop_at_entry: Optional[bool] = False,
+           startup_files: Optional[List[str]] = None) -> None:
+    """ Attach to given process and start a LLDB shell """
+    if not argv:
+        argv = None
+    if not envp:
+        envp = None
+    launch_lldb.launch(exec_path, argv, envp, stdin, stdout, stderr, wd, flags, stop_at_entry, startup_files)
+
+
```
Comment (doronz88) [inline hilda/cli.py:77]: let's add a `version` subcommand

File: hilda/lldb_importer.py [hunk 21]
```diff
@@ -0,0 +1,33 @@
+import subprocess
```
Comment (doronz88) [inline hilda/lldb_importer.py:1]: assuming we only operate from within `xcrun lldb`, is this still necessary?

File: tests/conftest.py [hunk 22]
```diff
@@ -1,7 +1,26 @@
...
+
+@pytest.fixture(scope='session')
+def lldb_debugger(request):
+    lldb_t = None
+    try:
+        lldb_t = LLDBAttachName(PROCESS)
+        lldb_t.start()
+        yield lldb_t.debugger
+    except LLDBException as e:
+        pytest.exit(f'{e.message} - Try use sudo')
```
Comment (doronz88) [inline tests/conftest.py:19]: let's catch a more specific exception for that

File: .github/workflows/python-app.yml [hunk 23]
```diff
@@ -17,26 +17,25 @@ jobs:
     strategy:
       matrix:
         python-version: [ 3.8, 3.9, "3.10", "3.11", "3.12" ]
-        os: [ ubuntu-latest, macos-latest, windows-latest ]
+        os: [ macos-latest ]
```
Comment (doronz88) [inline .github/workflows/python-app.yml:20]: let's also update the README & `pyproject.toml` accordingly

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: hilda/ipython_extensions/events.py [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(self, ip):
+        self.shell = ip
+        self.hilda_client = self.shell.user_ns['p']
+
+    def pre_run_cell(self, info):
+        """
+        Enable lazy loading for symbols
+        :param info: IPython's CellInfo object
+        """
+        if info.raw_cell[0] in ['!', '%'] or info.raw_cell.endswith('?'):
```
Comment (doronz88) [inline hilda/ipython_extensions/events.py:20]: stop is needed for that. avoid this entirely if it isn't, write a warning instead

File: hilda/hilda_client.py [hunk 25]
```diff
@@ -56,41 +57,66 @@
...
+        Whether to ignore breakpoints during evaluation.
+    nsobject_exclusion : bool
+        Whether to exclude NSObject during evaluation - reduce ipython autocomplete results.
+    objc_verbose_monitor : bool
+        When set to True, using monitor() will automatically print objc methods arguments.
+    """
+    evaluation_unwind_on_error: bool = False
+    evaluation_ignore_breakpoints: bool = False
+    nsobject_exclusion: bool = False
+    objc_verbose_monitor: bool = False
```
Comment (doronz88) [inline hilda/hilda_client.py:103]: add a `__repr__` and `__str__` explaining each of these values

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: hilda/ipython_extensions/keybindings.py (hunk 12 above)
Comment (netanelc305) [reply hilda/ipython_extensions/keybindings.py:10]: Changed it to access `ipython.user_ns['p']` exception will raise if no exsists

File: tests/conftest.py (hunk 22 above)
Comment (netanelc305) [reply tests/conftest.py:19]: discussed

File: hilda/launch_lldb.py (hunk 16 above)
Comment (doronz88) [reply hilda/launch_lldb.py]: let's also change it to an `if not` and `continue`

File: hilda/launch_lldb.py [hunk 26]
```diff
@@ -1,85 +1,192 @@
...
+    def __init__(self, address: str, port: int = 1234):
+        self.url_connect = f'connect://{address}:{port}'
+        super().__init__()
+
+    def _create_target(self) -> lldb.SBTarget:
+        return self.debugger.CreateTarget('')
+
+    def _create_process(self) -> lldb.SBProcess:
+        logger.debug(f'Connecting to "{self.url_connect}"')
+        process = self.target.ConnectRemote(self.listener, self.url_connect, None, self.error)
```
Comment (doronz88) [inline hilda/launch_lldb.py:88]: return here

File: hilda/launch_lldb.py [hunk 27]
```diff
@@ -1,85 +1,192 @@
...
+    def __init__(self, pid: int):
+        self.pid = pid
+        super().__init__()
+
+    def _create_target(self) -> lldb.SBTarget:
+        return self.debugger.CreateTargetWithFileAndArch(None, None)
+
+    def _create_process(self) -> lldb.SBProcess:
+        logger.debug(f'Attaching to {self.pid}')
+        process = self.target.AttachToProcessWithID(self.listener, self.pid, self.error)
```
Comment (doronz88) [inline hilda/launch_lldb.py:103]: return here

File: hilda/launch_lldb.py [hunk 28]
```diff
@@ -1,85 +1,192 @@
...
+        self.proc_name = proc_name
+        self.wait_for = wait_for
+        super().__init__()
+
+    def _create_target(self) -> lldb.SBTarget:
+        return self.debugger.CreateTargetWithFileAndArch(None, None)
+
+    def _create_process(self) -> lldb.SBProcess:
+        logger.debug(f'Attaching to {self.name}')
+        process = self.target.AttachToProcessWithName(self.listener, self.proc_name, self.wait_for, self.error)
```
Comment (doronz88) [inline hilda/launch_lldb.py:119]: return here

File: hilda/launch_lldb.py [hunk 29]
```diff
@@ -1,85 +1,192 @@
...
+
+    def _create_target(self) -> lldb.SBTarget:
+        return self.debugger.CreateTargetWithFileAndArch(self.exec_path, lldb.LLDB_ARCH_DEFAULT)
+
+    def _create_process(self) -> lldb.SBProcess:
+        # Launch(SBTarget self, SBListener listener, char const ** argv, char const ** envp,
+        # char const * stdin_path, char const * stdout_path, char const * stderr_path, char const * working_directory,
+        # uint32_t launch_flags, bool stop_at_entry, SBError error) -> SBProcess
+        logger.debug(f'Lunching process  {self.exec_path}')
+        process = self.target.Launch(self.listener, self.argv, self.envp,
```
Comment (doronz88) [inline hilda/launch_lldb.py:147]: return here

File: hilda/lldb_importer.py [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+import sys
+
+
+def get_lldb_python_path():
+    try:
+        # Run the lldb executable to print the Python path
+        result = subprocess.run(['lldb', '-P'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=True)
+        lldb_python_path = result.stdout.strip()
```
Comment (doronz88) [inline hilda/lldb_importer.py:9]: return here

File: hilda/lldb_importer.py [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+        return lldb_python_path
+    except subprocess.CalledProcessError as e:
+        print(f'Error running lldb: {e}')
```
Comment (doronz88) [inline hilda/lldb_importer.py:12]: shouldn't we raise this instead?

File: hilda/lldb_importer.py [hunk 32]
```diff
@@ -0,0 +1,32 @@
+import subprocess
+import sys
+
+
+def get_lldb_python_path():
```
Comment (doronz88) [inline hilda/lldb_importer.py:5]: add typing

File: hilda/ipython_extensions/events.py (hunk 24 above)
Comment (doronz88) [reply hilda/ipython_extensions/events.py:24]: TODO comment still remains

File: hilda/ipython_extensions/events.py [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+                    pass
+                else:
+                    try:
+                        self.hilda_client._add_global(
+                            node.id,
+                            symbol if symbol.type_ != lldb.eSymbolTypeObjCMetaClass else self.hilda_client.objc_get_class(
+                                node.id)
+                        )
+                    except EvaluatingExpressionError:
+                        self.hilda_client.log_warning('Process is running. Pause execution and try again')
```
Comment (doronz88) [inline hilda/ipython_extensions/events.py:49]:
```suggestion
                        self.hilda_client.log_warning(f'Process is running. Pause execution in order to resolve "{node.id}"')
```

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: hilda/lldb_importer.py (hunk 31 above)
Comment (doronz88) [reply hilda/lldb_importer.py]: raise where discussed

**Review by doronz88 — APPROVED** [review]

**Pull request merged**